### PR TITLE
Allow user to set col_atom_idxs for NonbondedInteractionGroup

### PR DESCRIPTION
**Pull request by jkausrelay**
- Update NonbondedInteractionGroup to accept col_atom_idxs
- If not set, default to current behavior of using all non-row idxs as the col idxs
- Update tests

### CODE REVIEW

File: timemachine/cpp/src/wrap_kernels.cpp
Comment (proteneer) [inline timemachine/cpp/src/wrap_kernels.cpp:1084]: maybe just break API and rename `atom_idxs` to `row_atom_idxs`?

Comment (proteneer) [reply timemachine/cpp/src/wrap_kernels.cpp:1084]: I'm actually in favor of `row` and `col` here still, because the underlying implementation is literally a square symmetric `NxN` matrix with distinct `rows` and `cols`, everything internally (including neighborlists), explicitly refer to `rows` and `cols` with different semantics, eg. `row` idxs are contiguous in ordering and `col` idxs. We've never used `group_1` or `group_2` in any other context before, and I'd be hesitant to introduce new notation.

Comment (proteneer) [reply timemachine/cpp/src/wrap_kernels.cpp:1084]: Anyways - I can see pros and cons of each @jkausrelay I'll let you decide.

Comment (jkausrelay) [reply timemachine/cpp/src/wrap_kernels.cpp:1084]: - Decided to go with row/col as that is consistent with the rest of the code (NC/NR, etc)

File: timemachine/cpp/src/nonbonded_interaction_group.hpp
Comment (mcwitt) [inline timemachine/cpp/src/nonbonded_interaction_group.hpp:73]: Should this file `#include <optional>`? (I get compilation errors locally without this)

File: timemachine/potentials/potentials.py
Comment (mcwitt) [inline timemachine/potentials/potentials.py:174]: nit: could this logic be moved to the reference potential implementation?

File: timemachine/cpp/generate_stubs
Comment (mcwitt) [inline timemachine/cpp/generate_stubs:11]: Nice workaround! Unfortunate that this seems necessary.

File: timemachine/cpp/src/nonbonded_interaction_group.cu
Comment (mcwitt) [inline timemachine/cpp/src/nonbonded_interaction_group.cu:57]: nit: I think in other places we `h` (or `d`) is a prefix instead of a suffix?

File: timemachine/cpp/src/nonbonded_interaction_group.cu
Comment (mcwitt) [inline timemachine/cpp/src/nonbonded_interaction_group.cu:244]: Seems like it might be possible to avoid the overhead of an extra kernel launch by making `d_row_indices` and `d_col_indices` contiguous. But from what I understand the overhead is pretty small (~20 us if I'm reading [this](https://developer.nvidia.com/blog/understanding-the-visualization-of-overhead-and-latency-in-nsight-systems/) correctly), so it may not be worth it. @badisa or @proteneer might know better.

File: timemachine/cpp/src/nonbonded_interaction_group.cu
Comment (mcwitt) [inline timemachine/cpp/src/nonbonded_interaction_group.cu:240]: nit: the naming of `K_B` seems a bit confusing to me. It's really a number of blocks, right? (would `B_K` make more sense?)

File: timemachine/cpp/src/nonbonded_interaction_group.cu
Comment (badisa) [reply timemachine/cpp/src/nonbonded_interaction_group.cu:249]: Would be nice to avoid this, though as @mcwitt said its not a big deal.

We could put the row indices at the start of an array and the column in the rest of array and use NC/NR to point at them specifically. Since the number of total indices can't exceed N. Then you could have a single call here. Would require re-working more of the potential.

Though not sure I entirely see a case where these two kernel calls produces significantly fewer NBlist rebuilds, so using ` k_check_rebuild_coords_and_box`  might also be reasonable.

File: timemachine/cpp/src/nonbonded_interaction_group.hpp
Comment (badisa) [reply timemachine/cpp/src/nonbonded_interaction_group.hpp:80]: I would argue for `vectors` in the C++. The set behavior is really for validation of indices which is handled by `verify_atom_indices`

File: tests/nonbonded/test_consistency.py
Comment (proteneer) [inline tests/nonbonded/test_consistency.py:57]: for my own sanity, maybe rename to `all_col_atom_idxs` to indicate that its a list of `col_atom_idxs`

File: tests/nonbonded/test_nonbonded_interaction_group.py
Comment (mcwitt) [reply tests/nonbonded/test_nonbonded_interaction_group.py:33]: Ah sorry, I meant to request a test for `row_atom_idxs`, `col_atom_idxs` that are different but overlapping (e.g. have one common index). I didn't see that this is tested?

File: timemachine/cpp/generate_stubs
Comment (proteneer) [reply timemachine/cpp/generate_stubs:11]: I have no idea what's going on here

File: timemachine/cpp/src/nonbonded_interaction_group.cu
Comment (proteneer) [inline timemachine/cpp/src/nonbonded_interaction_group.cu:29]: does `std::optional<std::set<int> >col_atom_idxs` overload the boolean operator so you can evaluate it like this? This is starting to hurt my brain.

File: timemachine/cpp/generate_stubs
Comment (mcwitt) [reply timemachine/cpp/generate_stubs:11]: I think this is to deal with the fact that mypy `stubgen` doesn't seem to add the required imports to `.pyi` files. (Might be unnecessary if we change the type of `col_atom_idxs` from `set` to `vector`?)

File: timemachine/cpp/src/nonbonded_interaction_group.hpp
Comment (proteneer) [inline timemachine/cpp/src/nonbonded_interaction_group.hpp:18]: Can we turn this into a method? i.e. `get_num_interacting_atoms()` - generally we'd like to avoid creating another member that can be computed from two other members.

Comment (proteneer) [reply timemachine/cpp/src/nonbonded_interaction_group.hpp:18]: (especially since `NR_` and `NC_` can change I think? and we'd like to avoid bookkeeping of keeping it synced s.t. `K_=NR_+NC_` for all values)

File: timemachine/cpp/src/nonbonded_interaction_group.hpp
Comment (proteneer) [inline timemachine/cpp/src/nonbonded_interaction_group.hpp:81]: I really don't want an `optional` for something this low-level, I'd rather the signature just be:
``` c++
NonbondedInteractionGroup(
        const int N,
        const std::vector<int> &row_atom_idxs,
        const std::vector<int> &col_atom_idxs,
        const double beta,
        const double cutoff
  )```

File: tests/nonbonded/test_nonbonded_interaction_group.py
Comment (proteneer) [inline tests/nonbonded/test_nonbonded_interaction_group.py:68]: What does -1 here do?

Comment (proteneer) [inline tests/nonbonded/test_nonbonded_interaction_group.py:68]: Add a test for num_atoms > 32 just to test for multi block / multi warp correctness

File: timemachine/cpp/src/nonbonded_interaction_group.cu
Comment (jkausrelay) [reply timemachine/cpp/src/nonbonded_interaction_group.cu:29]: Yes it does, but removed in the next commit

File: timemachine/cpp/generate_stubs
Comment (jkausrelay) [reply timemachine/cpp/generate_stubs]: Right. This was needed due to a missing import, removed since we are using vector now.

File: timemachine/cpp/src/local_md_utils.cu
Comment (badisa) [inline timemachine/cpp/src/local_md_utils.cu:65]: No longer used

File: timemachine/cpp/src/local_md_utils.cu
Comment (badisa) [inline timemachine/cpp/src/local_md_utils.cu:69]: It looks like this should fail due to this check
https://github.com/proteneer/timemachine/blob/split_ff_solv_ixn_gpu/timemachine/cpp/src/nonbonded_interaction_group.cu#L63-L68

File: timemachine/cpp/src/nonbonded_interaction_group.cu
Comment (badisa) [inline timemachine/cpp/src/nonbonded_interaction_group.cu:45]: For Local MD this change of `NR_ + NC_` will likely cause failures, as it reuse the potential with different sizes of indices, I would probably keep it `N_`. Otherwise, just need to modify https://github.com/proteneer/timemachine/blob/split_ff_solv_ixn_gpu/timemachine/cpp/src/local_md_utils.cu#L63 to construct ixn group that has size N


```suggestion
      beta_(beta), cutoff_(cutoff), nblist_(N_), nblist_padding_(nblist_padding), d_sort_storage_(nullptr),
```

**Review by proteneer — APPROVED** [review]
overall lgtm, pending tests

File: timemachine/cpp/src/nonbonded_interaction_group.cu
Comment (jkausrelay) [inline timemachine/cpp/src/nonbonded_interaction_group.cu:122]: @badisa I also changed this to use N_, otherwise this would overflow w/local MD when initialized only with the single dummy row/col idxs.

File: timemachine/cpp/src/local_md_utils.cu
Comment (badisa) [inline timemachine/cpp/src/local_md_utils.cu:6]: nit: Unused import

File: timemachine/cpp/src/nonbonded_interaction_group.cu
Comment (badisa) [inline timemachine/cpp/src/nonbonded_interaction_group.cu:371]: Don't do the same validation here that the constructor does, to make sure they are disjoint.

Also doesn't seem to ensure that the column indices are unique.

Comment (jkausrelay) [reply timemachine/cpp/src/nonbonded_interaction_group.cu:351]: Updated in https://github.com/proteneer/timemachine/pull/1011/commits/4c1a68a05f88abc2836f5719ba2a28c8db0db076 to use the same code for validation.

Fo local MD, is it correct to say the initializer should have non-zero NR/NC but set_atom_idxs can have 0 NR or NC? Want to make sure to check the right things here.

Comment (badisa) [reply timemachine/cpp/src/nonbonded_interaction_group.cu:351]: Correct. It is an inconsistency, but for Local MD it is an expectation that it can handle at least N particles, but can also be run on nothing.

**Review by badisa — APPROVED** [review]
🚀

**Review by mcwitt — APPROVED** [review]

**Pull request merged**
